pymchelper/readers/topas.py

```diff
@@ -0,0 +1,272 @@
+                                    'unit': match.group(3)}
+        else:
+            return None
+    return bins_data
+
+
+def extract_scorer_name(header_line: str) -> Optional[str]:
+    """Get scorer name from the output file"""
+    pattern = r"# Results for scorer: (\w+)"
+    match = re.search(pattern, header_line)
+    name = None
+    if match:
+        name = match.group(1)
+    return name
+
+
+def extract_scorer_unit_results(header_line: str) -> Optional[Tuple[str, str, List]]:
+    """Get scoring quantity, unit and the scoring values (sum/mean/etc.) from the output file"""
+    scorers = ['DoseToMedium', 'DoseToWater', 'DoseToMaterial', 'TrackLengthEstimator',
+               'AmbientDoseEquivalent', 'EnergyDeposit', 'Fluence', 'EnergyFluence',
+               'StepCount', 'OpticalPhotonCount', 'OriginCount', 'Charge', 'EffectiveCharge',
+               'ProtonLET', 'SurfaceCurrent', 'SurfaceTrackCount', 'PhaseSpace']
+    for scorer in scorers:
+        if scorer in header_line:
+            unit = ""
+            pattern = f"# {scorer}\\s(\\( .*? \\))?\\s*: (.*)"
+            match = re.search(pattern, header_line)
+            if match:
+                unit = ""
+                if match.group(1):
+                    unit = match.group(1)[2:-2]
+                results = match.group(2).split()
+                return (scorer, unit, results)
+    return None
+
+
+def extract_differential_axis(header_line: str) -> Optional[MeshAxis]:
+    """Check if the output file contains differential axis and get it from file if it does"""
+    if "# Binned by" in header_line:
+        pattern = r"# Binned by (.+?) in (\d+) bin[s ] of (\d+) (\w+) from ([\d.]+) (\w+) to ([\d.]+) (\w+)"
+        match = re.search(pattern, header_line)
+        if match:
+            binned_by = match.group(1)
+            num_bins = int(match.group(2))
+            unit = match.group(4)
+            min_val = float(match.group(5))
+            max_val = float(match.group(7))
+
+            return MeshAxis(n=num_bins, min_val=min_val, max_val=max_val,
+                            name=binned_by, unit=unit, binning=MeshAxis.BinningType.linear)
+    return None
+
+
+class TopasReader(Reader):
+    """Reader for Topas output files"""
+
+    def __init__(self, filename):
+        super(TopasReader, self).__init__(filename)
+        self.directory = Path(filename).parent
+
+    def read_data(self, estimator: Estimator) -> bool:
+        """
+        Read the data from the file and store them in the provided estimator object.
+        Topas reader assumes that the input file is in the same directory as the output file
+        to extract the number of histories from it.
+        If the name of the input file is not found in the output file
+        or the input file is not in the same directory, the number of histories is set to 0.
+        """
+        with open(self.filename, 'r') as results_file:
+            results_lines = results_file.readlines()
```

Comment:
I don't like the programming style with `break`.
What do you think of creating header lines or comment lines:

```comment_lines = [line for results_lines if line.startswith("#")]```

and
```data_lines = [line for results_lines if not line.startswith("#")]```

? Then you could adjust the code.

Reply:
Yes, I think it looks better